File: src/dynamic_steering_aligner.cpp
// ==========================================
// Basic Libraries & Headers
// ==========================================
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/float32.hpp>
#include <std_msgs/msg/float32_multi_array.hpp>
#include <std_msgs/msg/bool.hpp>
#include <rclcpp/qos.hpp>
#include "heading_controller_pkg/dynamic_steering_aligner_utils.hpp"

// ==========================================
// Type Aliases for Standard ROS 2 Messages
// ==========================================
using Float32 = std_msgs::msg::Float32;
using Float32MultiArray = std_msgs::msg::Float32MultiArray;
using Bool = std_msgs::msg::Bool;

// ==========================================
// Steering Aligner Node Class
// Adaptive proportional gain + deadband + EMA smoothing + slew-rate
// limiting + oscillation detection, with a gentler hold/stabilize phase
// once the target heading is reached.
// (angle_diff() / adaptive_gain() / apply_ema() / slew_limit() /
// update_sign_history() / detect_oscillation() / compute_aligner_step()
// now live in dynamic_steering_aligner_utils.hpp so they can be
// unit-tested directly with gtest, without needing a running rclcpp::Node)
//
// NOTE: min_pwm is accepted as a parameter for parity with the Python
// node, but -- exactly like the Python -- it is NOT used for clamping;
// the clamp uses -abs(max_pwm) instead. Preserved faithfully.
// ==========================================
class SteeringAligner : public rclcpp::Node {
public:
    SteeringAligner() : Node("steering_aligner") {
        // ----------------------------------------------------
        // Declare ROS 2 Parameters (defaults match the Python node)
        // ----------------------------------------------------
        this->declare_parameter<float>("tolerance_deg", 2.0f);
        this->declare_parameter<float>("deadband_deg", 0.25f);
        this->declare_parameter<float>("max_pwm", 17.0f);
        this->declare_parameter<float>("min_pwm", -17.0f);
        this->declare_parameter<float>("k_min", 3.08f);
        this->declare_parameter<float>("k_max", 0.9f);
        this->declare_parameter<float>("gain_scale", 6.0f);
        this->declare_parameter<float>("ema_alpha_base", 0.25f);
        this->declare_parameter<float>("max_slew_rate", 80.0f);
        this->declare_parameter<float>("hold_duration", 5.0f);
        this->declare_parameter<float>("hold_k_max_factor", 1.1f);
        this->declare_parameter<float>("hold_ema_alpha_factor", 0.95f);
        this->declare_parameter<float>("osc_window_sec", 1.2f);
        this->declare_parameter<int>("osc_sign_changes_threshold", 5);
        this->declare_parameter<float>("osc_reduction_factor", 0.45f);
        this->declare_parameter<float>("osc_min_time_between_warnings", 1.5f);
        this->declare_parameter<float>("disable_time", 0.0f);
        this->declare_parameter<float>("loop_hz", 100.0f);

        // ----------------------------------------------------
        // Load Parameters into Member Variables
        // ----------------------------------------------------
        params_.tolerance_deg = static_cast<float>(this->get_parameter("tolerance_deg").as_double());
        params_.deadband_deg = static_cast<float>(this->get_parameter("deadband_deg").as_double());
        params_.max_pwm = static_cast<float>(this->get_parameter("max_pwm").as_double());
        params_.min_pwm = static_cast<float>(this->get_parameter("min_pwm").as_double());
        params_.k_min = static_cast<float>(this->get_parameter("k_min").as_double());
        params_.k_max = static_cast<float>(this->get_parameter("k_max").as_double());
        params_.gain_scale = static_cast<float>(this->get_parameter("gain_scale").as_double());
        params_.ema_alpha_base = static_cast<float>(this->get_parameter("ema_alpha_base").as_double());
        params_.max_slew_rate = static_cast<float>(this->get_parameter("max_slew_rate").as_double());
        params_.hold_duration = static_cast<float>(this->get_parameter("hold_duration").as_double());
        params_.hold_k_max_factor = static_cast<float>(this->get_parameter("hold_k_max_factor").as_double());
        params_.hold_ema_alpha_factor = static_cast<float>(this->get_parameter("hold_ema_alpha_factor").as_double());
        params_.osc_window_sec = static_cast<double>(this->get_parameter("osc_window_sec").as_double());
        params_.osc_sign_changes_threshold = this->get_parameter("osc_sign_changes_threshold").as_int();
        params_.osc_reduction_factor = static_cast<float>(this->get_parameter("osc_reduction_factor").as_double());
        params_.osc_min_time_between_warnings = static_cast<double>(this->get_parameter("osc_min_time_between_warnings").as_double());
        params_.disable_time = static_cast<double>(this->get_parameter("disable_time").as_double());
        float loop_hz = static_cast<float>(this->get_parameter("loop_hz").as_double());

        // ----------------------------------------------------
        // Setup Publishers & Subscribers
        // ----------------------------------------------------
        rclcpp::QoS qos_policy(1);
        qos_policy.reliability(rclcpp::ReliabilityPolicy::Reliable);
        qos_policy.durability(rclcpp::DurabilityPolicy::Volatile);

        steer_pub_ = this->create_publisher<Float32>("/steering_angle", qos_policy);
        align_done_pub_ = this->create_publisher<Bool>("/alignment_done", qos_policy);

        aligned_sub_ = this->create_subscription<Float32>(
            "/aligned", 10, std::bind(&SteeringAligner::aligned_callback, this, std::placeholders::_1));
        compass_sub_ = this->create_subscription<Float32MultiArray>(
            "/jmoab_compass", 10, std::bind(&SteeringAligner::compass_callback, this, std::placeholders::_1));

        // ----------------------------------------------------
        // Initialize State
        // ----------------------------------------------------
        state_ = AlignerState{};
        state_.last_cmd_time = this->now().seconds();

        // ----------------------------------------------------
        // Setup Control Loop Timer
        // ----------------------------------------------------
        timer_ = this->create_wall_timer(
            std::chrono::duration<float>(1.0f / std::max(1.0f, loop_hz)),
            std::bind(&SteeringAligner::control_loop, this));

        RCLCPP_INFO(this->get_logger(),
                    "🧭 Steering Aligner Node started (adaptive, deadband, slew, EMA, osc-detector).");
    }

private:
    // ==========================================
    // Callbacks
    // ==========================================
    void aligned_callback(const Float32::SharedPtr msg) {
        double now = this->now().seconds();
        if (now - state_.last_disable_time < params_.disable_time) {
            return;
        }
        state_.target_heading = wrap360(msg->data);
        state_.active = true;
        state_.holding = false;
        RCLCPP_INFO(this->get_logger(), "🎯 New target heading: %.2f°", state_.target_heading.value());
    }

    void compass_callback(const Float32MultiArray::SharedPtr msg) {
        if (msg->data.size() >= 3) {
            state_.current_heading = wrap360(msg->data[2]);
        }
    }

    // ==========================================
    // Main control loop
    // ==========================================
    void control_loop() {
        double now = this->now().seconds();
        AlignerStepResult result = compute_aligner_step(state_, now, params_);

        if (result.oscillation_warning) {
            RCLCPP_WARN(this->get_logger(), "⚠️ Oscillation detected — reducing aggressiveness temporarily.");
        }

        if (result.entered_hold) {
            RCLCPP_INFO(this->get_logger(), "✅ Aligned within ±%.2f°. Entering hold for %.1fs.",
                        params_.tolerance_deg, params_.hold_duration);
            return;
        }

        if (result.should_publish_pwm) {
            publish_pwm(result.pwm_value);
            return;
        }

        if (result.alignment_done) {
            publish_alignment_done();
            RCLCPP_INFO(this->get_logger(), "📡 Alignment complete and stabilized — released control.");
            return;
        }
    }

    // ==========================================
    // Publishers
    // ==========================================
    void publish_pwm(float pwm_value) {
        Float32 msg;
        msg.data = pwm_value;
        steer_pub_->publish(msg);
    }

    void publish_alignment_done() {
        Bool msg;
        msg.data = true;
        align_done_pub_->publish(msg);
        RCLCPP_INFO(this->get_logger(), "📡 Published /alignment_done = True");
    }

    // ==========================================
    // ROS 2 Communication Handles
    // ==========================================
    rclcpp::Publisher<Float32>::SharedPtr steer_pub_;
    rclcpp::Publisher<Bool>::SharedPtr align_done_pub_;
    rclcpp::Subscription<Float32>::SharedPtr aligned_sub_;
    rclcpp::Subscription<Float32MultiArray>::SharedPtr compass_sub_;
    rclcpp::TimerBase::SharedPtr timer_;

    // ==========================================
    // Configuration Parameters
    // ==========================================
    AlignerParams params_;

    // ==========================================
    // Internal State
    // ==========================================
    AlignerState state_;
};

// ==========================================
// Main Execution Entry Point
// ==========================================
int main(int argc, char** argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<SteeringAligner>();

    try {
        rclcpp::spin(node);
    } catch (const std::exception& e) {
        RCLCPP_INFO(node->get_logger(), "🛑 Steering Aligner interrupted by user.");
    }

    rclcpp::shutdown();
    return 0;
}

File: python_files/dynamic_steering_aligner.py
#!/usr/bin/env python3
"""
Adaptive & Robust Steering Aligner Node — ROS2 Humble

Features:
 - Adaptive gain: stronger correction for large errors, gentle for small ones.
 - Deadband: avoids micro-corrections and hunting.
 - EMA smoothing of published steering command.
 - Slew-rate limiter to avoid abrupt jumps (prevents overshoot).
 - Oscillation detector: if repeated sign changes occur, temporarily reduce aggressiveness.
 - Separate, gentler behavior during HOLD/STABILIZE phase.

Topics:
 - Subscribes:
     /aligned (std_msgs/Float32)           -> target heading [deg]
     /jmoab_compass (std_msgs/Float32MultiArray) -> [roll,pitch,heading]
 - Publishes:
     /steering_angle (std_msgs/Float32)
     /alignment_done (std_msgs/Bool)
"""
import rclpy
from rclpy.node import Node
from std_msgs.msg import Float32, Float32MultiArray, Bool
import math
import time
from collections import deque
from rclpy.qos import QoSProfile, ReliabilityPolicy, QoSDurabilityPolicy


class SteeringAligner(Node):
    def __init__(self):
        super().__init__('steering_aligner')

        # ---------------------------
        # TUNABLE PARAMETERS (start here)
        # ---------------------------
        # Alignment tolerance (deg). Within ± this -> start hold phase.
        self.tolerance_deg = 2.0

        # Deadband (deg): errors inside this are considered zero (reduces hunting).
        self.deadband_deg = 0.25

        # PWM output limits (signed).
        self.max_pwm = 17.0
        self.min_pwm = -17.0

        # Adaptive gain shaping:
        self.k_min = 3.08    # gentle proportional scale for very small errors
        self.k_max = 0.9     # maximum proportional scale for large errors
        self.gain_scale = 6.0  # degrees over which gain rises toward k_max (larger -> slower ramp)

        # EMA smoothing for command: base alpha (0..1). Lower = more smoothing.
        self.ema_alpha_base = 0.25

        # Slew-rate limit (PWM units per second)
        self.max_slew_rate = 80.0

        # Hold / stabilization behavior
        self.hold_duration = 5.0
        self.hold_k_max_factor = 1.1  # fraction of k_max used during hold (much gentler)
        self.hold_ema_alpha_factor = 0.95  # stronger smoothing in hold

        # Oscillation detection parameters
        self.osc_window_sec = 1.2
        self.osc_sign_changes_threshold = 5
        self.osc_reduction_factor = 0.45  # reduce k_max by this factor while oscillating
        self.osc_min_time_between_warnings = 1.5

        # Timer frequency
        self.loop_hz = 100.0  # 100 Hz control loop

        # Cooldown after alignment (seconds)
        self.disable_time = 0.0

        # ---------------------------
        # Internal state
        # ---------------------------
        self.target_heading = None
        self.current_heading = None
        self.smoothed_pwm = 0.0
        self.last_pwm_cmd = 0.0
        self.last_cmd_time = time.time()
        self.active = True
        self.holding = False
        self.hold_start_time = None
        self.last_disable_time = 0.0

        # For oscillation detection (deque of (timestamp, sign))
        self.sign_history = deque()
        self.last_osc_warn_time = 0.0

        # ---------------------------
        # Publishers & Subscribers
        # ---------------------------
        qos_policy = QoSProfile(
            reliability=ReliabilityPolicy.RELIABLE,
            durability=QoSDurabilityPolicy.VOLATILE,
            depth=1
        )
        self.steer_pub = self.create_publisher(Float32, '/steering_angle', qos_policy)
        self.align_done_pub = self.create_publisher(Bool, '/alignment_done', qos_policy)

        self.create_subscription(Float32, '/aligned', self.aligned_callback, 10)
        self.create_subscription(Float32MultiArray, '/jmoab_compass', self.compass_callback, 10)

        self.create_timer(1.0 / self.loop_hz, self.control_loop)
        self.get_logger().info("🧭 Steering Aligner Node started (adaptive, deadband, slew, EMA, osc-detector).")

    # ---------------------------
    # Callbacks
    # ---------------------------
    def aligned_callback(self, msg: Float32):
        """New target heading in degrees (0..360)."""
        if time.time() - self.last_disable_time < self.disable_time:
            return
        self.target_heading = msg.data % 360.0
        self.active = True
        self.holding = False
        self.get_logger().info(f"🎯 New target heading: {self.target_heading:.2f}°")

    def compass_callback(self, msg: Float32MultiArray):
        """Expecting [roll, pitch, heading]."""
        if len(msg.data) >= 3:
            self.current_heading = msg.data[2] % 360.0

    # ---------------------------
    # Utility helpers
    # ---------------------------
    @staticmethod
    def angle_diff(a, b):
        """
        Smallest signed difference between headings a and b (degrees).
        Return in range (-180, +180].
        """
        d = (a - b + 180.0) % 360.0 - 180.0
        return d

    def apply_deadband(self, value, deadband_deg):
        """Convert deadband in degrees to approximate pwm deadband using k_min and scaling.
           But here we use direct degree deadband: if error_abs < deadband -> zero."""
        return 0.0 if abs(value) <= deadband_deg else value

    def adaptive_gain(self, error_abs, k_max_effective):
        """Exponential approach mapping |error| -> gain in [k_min, k_max_effective]."""
        # 1 - exp(-e/gain_scale) gives a smooth curve
        return self.k_min + (k_max_effective - self.k_min) * (1.0 - math.exp(-error_abs / max(1e-6, self.gain_scale)))

    def apply_ema(self, new_val, alpha):
        """EMA smoothing of command value."""
        self.smoothed_pwm = (alpha * new_val) + (1.0 - alpha) * self.smoothed_pwm
        return self.smoothed_pwm

    def slew_limit(self, desired, last, dt):
        """Limit rate of change of PWM command."""
        if dt <= 0:
            return desired
        max_delta = self.max_slew_rate * dt
        delta = desired - last
        if abs(delta) > max_delta:
            desired = last + math.copysign(max_delta, delta)
        return desired

    def update_sign_history(self, error):
        """Append sign record and prune old entries."""
        now = time.time()
        if error > 0:
            sign = 1
        elif error < 0:
            sign = -1
        else:
            sign = 0
        self.sign_history.append((now, sign))
        cutoff = now - self.osc_window_sec
        while self.sign_history and self.sign_history[0][0] < cutoff:
            self.sign_history.popleft()

    def detect_oscillation(self):
        """Return True if sign changes in recent window exceed threshold."""
        signs = [s for (_, s) in self.sign_history if s != 0]
        if len(signs) < 2:
            return False
        changes = 0
        prev = signs[0]
        for s in signs[1:]:
            if s != prev:
                changes += 1
                prev = s
        return changes >= self.osc_sign_changes_threshold

    # ---------------------------
    # Main control loop
    # ---------------------------
    def control_loop(self):
        now = time.time()
        dt = max(1e-6, now - self.last_cmd_time)
        self.last_cmd_time = now

        if self.current_heading is None or self.target_heading is None:
            return

        if not self.active and not self.holding:
            return

        # Signed heading error (target - current), positive => need to turn "positive" direction
        error = self.angle_diff(self.target_heading, self.current_heading)
        error_abs = abs(error)

        # Update oscillation detector sign history & check
        self.update_sign_history(error)
        oscillating = self.detect_oscillation()

        # If detected oscillation, reduce k_max temporarily and increase smoothing
        k_max_effective = self.k_max
        ema_alpha = self.ema_alpha_base
        if oscillating:
            k_max_effective = max(self.k_min, self.k_max * self.osc_reduction_factor)
            ema_alpha = max(0.02, self.ema_alpha_base * 0.6)
            # throttle warning logs
            if now - self.last_osc_warn_time > self.osc_min_time_between_warnings:
                self.get_logger().warn("⚠️ Oscillation detected — reducing aggressiveness temporarily.")
                self.last_osc_warn_time = now

        # If in aligning phase
        if not self.holding:
            # If within tolerance -> enter hold/stabilize phase
            if error_abs <= self.tolerance_deg:
                self.holding = True
                self.hold_start_time = now
                self.get_logger().info(f"✅ Aligned within ±{self.tolerance_deg:.2f}°. Entering hold for {self.hold_duration}s.")
                # don't send immediate zero; let hold handle gentle corrections
                return

            # Compute adaptive proportional gain
            gain = self.adaptive_gain(error_abs, k_max_effective)

            # Proportional command: sign preserved
            pwm_cmd = gain * error_abs

            # Clip to allowed range
            pwm_cmd = max(-abs(self.max_pwm), min(self.max_pwm, pwm_cmd))
            pwm_cmd = pwm_cmd if error >= 0 else -pwm_cmd

            # Apply a deadband in degrees -> convert to minimal pwm threshold
            # Simpler: if error_abs < deadband_deg -> 0
            if error_abs <= self.deadband_deg:
                pwm_cmd = 0.0

            # Slew-rate limit (smooth sudden jumps)
            pwm_cmd = self.slew_limit(pwm_cmd, self.last_pwm_cmd, dt)

            # EMA smoothing
            pwm_cmd = self.apply_ema(pwm_cmd, ema_alpha)

            # Publish
            self.publish_pwm(pwm_cmd)
            self.last_pwm_cmd = pwm_cmd

        # Holding / stabilization phase: make much gentler corrections and stronger smoothing
        else:
            hold_elapsed = now - self.hold_start_time
            if hold_elapsed < self.hold_duration:
                # Use much smaller max gain and smaller commands
                hold_k_max = max(self.k_min, k_max_effective * self.hold_k_max_factor)
                gain = self.adaptive_gain(error_abs, hold_k_max)

                pwm_cmd = gain * error_abs
                pwm_cmd = max(-abs(self.max_pwm), min(self.max_pwm, pwm_cmd))
                pwm_cmd = pwm_cmd if error >= 0 else -pwm_cmd

                # Slightly larger deadband during hold to avoid new micro corrections
                if error_abs <= (self.deadband_deg * 1.5):
                    pwm_cmd = 0.0

                # More conservative slew & EMA in hold
                pwm_cmd = self.slew_limit(pwm_cmd, self.last_pwm_cmd, dt)
                pwm_cmd = self.apply_ema(pwm_cmd, max(0.01, ema_alpha * self.hold_ema_alpha_factor))

                self.publish_pwm(pwm_cmd)
                self.last_pwm_cmd = pwm_cmd
                return
            else:
                # Finished hold phase -> publish alignment done and release control
                self.publish_alignment_done()
                self.active = False
                self.holding = False
                self.last_disable_time = now
                self.get_logger().info("📡 Alignment complete and stabilized — released control.")
                return

    # ---------------------------
    # Publishers
    # ---------------------------
    def publish_pwm(self, pwm_value):
        msg = Float32()
        msg.data = float(pwm_value)
        self.steer_pub.publish(msg)

    def publish_alignment_done(self):
        msg = Bool()
        msg.data = True
        self.align_done_pub.publish(msg)
        self.get_logger().info("📡 Published /alignment_done = True")


def main(args=None):
    rclpy.init(args=args)
    node = SteeringAligner()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.get_logger().info("🛑 Steering Aligner interrupted by user.")
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

